kytos/core/link.py

```diff
@@ -121,65 +123,55 @@ def id(self):  # pylint: disable=invalid-name
+            self.endpoint_b.available_tags[tag_type],
+        )
 
-    def is_tag_available(self, tag):
+    def is_tag_available(self, tag: int, tag_type: str = 'vlan'):
         """Check if a tag is available."""
-        return (self.endpoint_a.is_tag_available(tag) and
-                self.endpoint_b.is_tag_available(tag))
-
-    def get_next_available_tag(self):
+        return (self.endpoint_a.is_tag_available(tag, tag_type) and
+                self.endpoint_b.is_tag_available(tag, tag_type))
+
+    def get_next_available_tag(
+        self,
+        controller,
+        tag_type: str = 'vlan'
+    ) -> int:
         """Return the next available tag if exists."""
         with self._get_available_vlans_lock:
-            # Copy the available tags because in case of error
-            # we will remove and add elements to the available_tags
-            available_tags_a = self.endpoint_a.available_tags.copy()
-            available_tags_b = self.endpoint_b.available_tags.copy()
-
-            for tag in available_tags_a:
-                # Tag does not exist in endpoint B. Try another tag.
-                if tag not in available_tags_b:
-                    continue
-
-                # Tag already in use. Try another tag.
-                if not self.endpoint_a.use_tag(tag):
-                    continue
-
-                # Tag already in use in B. Mark the tag as available again.
-                if not self.endpoint_b.use_tag(tag):
-                    self.endpoint_a.make_tag_available(tag)
-                    continue
-
-                # Tag used successfully by both endpoints. Returning.
-                return tag
+            with self.endpoint_a_lock:
+                with self.endpoint_b_lock:
+                    ava_tags_a = self.endpoint_a.available_tags[tag_type]
+                    ava_tags_b = self.endpoint_b.available_tags[tag_type]
+                    tag = Interface.range_intersection_first_tag(ava_tags_a,
+                                                                 ava_tags_b)
+                    if tag:
+                        self.endpoint_a.use_tags([tag, tag])
+                        self.endpoint_a.notify_interface_tags(controller)
+                        self.endpoint_b.use_tags([tag, tag])
+                        self.endpoint_b.notify_interface_tags(controller)
+                        return tag
 
             raise KytosNoTagAvailableError(self)
 
-    def make_tag_available(self, tag):
+    def make_tags_available(
+        self,
+        controller,
+        tag: int,
+        tag_type: str = 'vlan'
+    ) -> (bool, bool):
         """Add a specific tag in available_tags."""
-        if not self.is_tag_available(tag):
-            self.endpoint_a.make_tag_available(tag)
-            self.endpoint_b.make_tag_available(tag)
-            return True
-        return False
+        result_a = self.endpoint_a.make_tags_available([tag, tag], tag_type)
```

Comment:
Let's also acquire `_get_available_vlans_lock[link.id]` before all these operations just so releasing the tag here would be thread safe. Ideally you'll also need to acquire explicitly the `self.endpoint_a._tag_lock` and `self.endpoint_b._tag_lock` nested (same triple pattern we talked on get_next_available_tag), you'll also need to parametrize a `lock` parameter on `make_tags_available` otherwise you'll hit an issue of re-entrant lock. 

Essentially, the major point of the "triple nested lock" (link / intf_a / intf_b) is to ensure that once an operation starts with a link, it'll also in order acquire the interface locks from an outer scope, ensuring that individually none of these interfaces will change their tags.

Reply:
Added triple lock.

Reply:
Looks good.